syntax = "proto3";

package raft;

service RaftService {
  rpc AddLog (AddLogRequest) returns (AddLogResponse);
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse);
}

message RequestVoteRequest {
  int32 candidate_term = 1;
  int32 candidate_log_term = 2;
  int32 candidate_log_length = 3;
  string candidate_id = 4;
}

message RequestVoteResponse {
  string voter_id = 1;
  int32 term = 2;
  bool granted = 3;
}

message AddLogRequest {
  string leader_id = 1;
  int32 term = 2;
  int32 prefix_len = 3;
  int32 prefix_term = 4;
  int32 commit_length = 5;
  repeated pair suffix = 6;
  int32 leader_commit = 7;
}

message pair {
  string first = 1;
  int32 second = 2;
}

message AddLogResponse {
  string follower = 1;
  int32 term = 2;
  int32 ack = 3;
  bool success = 4;
}


service ClientService {
  rpc Write (WriteRequest) returns (WriteResponse);
  rpc Read (ReadRequest) returns (ReadResponse);
}

message WriteRequest {
  string long_url = 1;
}

message WriteResponse {
  string long_url = 1;
  string short_url = 2;
}

message ReadRequest {
  string url = 1; //can be both short or long url
}

message ReadResponse {
  string url = 1; //short or long url
  string result_url = 2; //long or short url
}